src/main/java/io/github/acm19/aws/interceptor/http/AwsRequestSigningApacheV5Interceptor.java

```diff
@@ -67,38 +77,71 @@ public AwsRequestSigningApacheV5Interceptor(String service,
      * {@inheritDoc}
      */
     @Override
-    public void process(HttpRequest request, EntityDetails entityDetails, HttpContext context)
-            throws HttpException, IOException {
+    public void execute(
+        HttpRequest request,
+        AsyncEntityProducer entityProducer,
+        AsyncExecChain.Scope scope,
+        AsyncExecChain chain,
+        AsyncExecCallback asyncExecCallback)
+        throws HttpException, IOException {
+        signRequest(request, getContentStreamSupplier(scope.originalRequest));
+        chain.proceed(request, entityProducer, scope, asyncExecCallback);
+    }
+
+    @Override
+    public ClassicHttpResponse execute(ClassicHttpRequest classicHttpRequest,
+                                       ExecChain.Scope scope,
+                                       ExecChain execChain) throws IOException, HttpException {
+        signRequest(classicHttpRequest, getContentStreamSupplier(classicHttpRequest));
+        return execChain.proceed(classicHttpRequest, scope);
+    }
+
+    private void signRequest(HttpRequest request, Supplier<InputStream> contentStreamSupplier) throws IOException {
         // copy Apache HttpRequest to AWS request
-        SdkHttpFullRequest.Builder requestBuilder = SdkHttpFullRequest.builder()
+        SdkHttpFullRequest.Builder requestBuilder =
+            SdkHttpFullRequest.builder()
                 .method(SdkHttpMethod.fromValue(request.getMethod()))
                 .uri(buildUri(request));
 
-        if (request instanceof ClassicHttpRequest) {
-            ClassicHttpRequest classicHttpRequest = (ClassicHttpRequest) request;
+        if (contentStreamSupplier != null) {
+            requestBuilder
+                .contentStreamProvider(ContentStreamProvider.fromInputStreamSupplier(contentStreamSupplier));
```

Comment:
Indent to 8 spaces, I'd actually break the parameter or statically import `fromInputStreamSupplier`.